Makefile

```diff
@@ -14,74 +25,298 @@ export BROWSER_PYSCRIPT
 
-servedocs: docs ## compile the docs watching for changes
-	watchmedo shell-command -p '*.rst' -c '$(MAKE) -C docs html' -R -D .
 
-release: clean ## package and upload a release
-	python setup.py sdist upload
-	python setup.py bdist_wheel upload
+#################################################################################
+## Meta Targets ##
+#################################################################################
 
-dist: clean ## builds source and wheel package
-	python setup.py sdist
-	python setup.py bdist_wheel
-	ls -l dist
+test-all: ensure-certs ## run full e2e tests across automated profiles (requires: make certs)
+	@echo "Running full e2e tests across automated profiles..."
+	@for profile in single-user secure-ldap secure-mtls; do \
+		echo "=== Running e2e test for profile: $$profile ==="; \
+		$(MAKE) up NIPYAPI_PROFILE=$$profile && \
+		$(MAKE) wait-ready NIPYAPI_PROFILE=$$profile && \
+		$(MAKE) test NIPYAPI_PROFILE=$$profile; \
+		test_result=$$?; \
+		if [ $$test_result -ne 0 ]; then \
+			echo "Tests failed for profile: $$profile"; \
+			exit $$test_result; \
+		fi; \
+	done
+	@echo "✅ All profiles tested successfully"
 
-install: clean ## install the package to the active Python's site-packages
-	python setup.py install
+sandbox: ensure-certs ## create isolated environment with sample objects: make sandbox NIPYAPI_PROFILE=single-user|secure-ldap|secure-mtls|secure-oidc
+	@if [ -z "$(NIPYAPI_PROFILE)" ]; then echo "❌ NIPYAPI_PROFILE is required (single-user|secure-ldap|secure-mtls|secure-oidc)"; exit 1; fi
+	@echo "🏗️ Setting up NiPyAPI sandbox with profile: $(NIPYAPI_PROFILE)"
+	@echo "=== 1/4: Starting infrastructure ==="
+	$(MAKE) up NIPYAPI_PROFILE=$(NIPYAPI_PROFILE)
+	@echo "=== 2/4: Waiting for readiness ==="
+	$(MAKE) wait-ready NIPYAPI_PROFILE=$(NIPYAPI_PROFILE)
+	@echo "=== 3/4: Setting up authentication and sample objects ==="
+	@NIPYAPI_PROFILE=$(NIPYAPI_PROFILE) python examples/sandbox.py $(NIPYAPI_PROFILE)
+
+rebuild-all: ## comprehensive rebuild: clean -> certs -> extract APIs -> gen clients -> test all -> build -> validate -> docs
+	@echo "🚀 Starting comprehensive rebuild from clean slate..."
+	@echo "=== 1/9: Clean All Artifacts ==="
+	$(MAKE) clean-all
+	@echo "=== 2/9: Generate Certificates ==="
+	$(MAKE) certs
+	@echo "=== 3/9: Extract OpenAPI Specs ==="
+	$(MAKE) up NIPYAPI_PROFILE=single-user
+	$(MAKE) wait-ready NIPYAPI_PROFILE=single-user
+	$(MAKE) fetch-openapi
+	@echo "=== 4/9: Augment OpenAPI Specs ==="
+	$(MAKE) augment-openapi
+	@echo "=== 5/9: Generate Fresh Clients ==="
+	$(MAKE) gen-clients
+	@echo "=== 6/9: Test All Profiles with Fresh Clients ==="
+	$(MAKE) test-all
+	@echo "=== 7/9: Build Distribution Packages ==="
+	$(MAKE) dist
+	@echo "=== 8/9: Validate Distribution Packages ==="
+	$(MAKE) check-dist
+	$(MAKE) test-dist
+	@echo "=== 9/9: Generate Documentation ==="
+	$(MAKE) docs
+	@echo "✅ Comprehensive rebuild completed successfully!"
```

Comment:
This is a lot of stuff, consider how it would work if you broke it down into multiple makefiles, such that you had a top level Makefile and _n_ included .mk files that had targets related by function.

It is a trade off between working in a big file and keeping the dispersed targets in mind

Reply:
It's ~330 lines at the moment, in 3 sections. If we need to expand it further then breaking it apart is a good idea, but for now it feels maintainable to me.